File: leetcode/102. Binary Tree Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(root == NULL){
            return res;
        }
        queue<TreeNode*> q;
        q.push(root);
        while(q.size() != 0){
            int size = q.size();
            vector<int> level;
            while(size != 0){
                TreeNode* node = q.front();
                q.pop();
                level.push_back(node->val);
                if(node->left != NULL){
                    q.push(node->left);
                }
                if(node->right != NULL){
                    q.push(node->right);
                }
                --size;
            }
            res.push_back(std::move(level));
        }
        return res;
    }
};

File: leetcode/102. Binary Tree Level Order Traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::collections::VecDeque;
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut res = Vec::new();
        if root == None {
            return res;
        }
        let mut q = VecDeque::new();
        q.push_back(root);
        
        unsafe {
            while q.len() != 0 {
                let mut size = q.len();
                let mut level = Vec::new();
                while size != 0 {
                    let node = q.pop_front().unwrap().unwrap();
                    level.push(node.borrow().val);
                    if node.borrow().left != None {
                        q.push_back(node.borrow().left.clone());
                    }
                    if node.borrow().right != None {
                        q.push_back(node.borrow().right.clone())
                    }
                    size -= 1;
                }
                res.push(level);
            }
        }
        
        res
    }
}
